pilot/pkg/model/jwks_resolver.go

```diff
@@ -236,48 +239,28 @@ func (r *JwksResolver) GetPublicKey(issuer string, jwksURI string) (string, erro
 		}
 		return e.pubKey, nil
 	}
-
-	var err error
-	var pubKey string
-	if jwksURI == "" {
-		// Fetch the jwks URI if it is not hardcoded on config.
-		jwksURI, err = r.resolveJwksURIUsingOpenID(issuer)
-	}
-	if err != nil {
-		log.Errorf("Failed to jwks URI from %q: %v", issuer, err)
-	} else {
-		var resp []byte
-		resp, err = r.getRemoteContentWithRetry(jwksURI, networkFetchRetryCountOnMainFlow)
-		if err != nil {
-			log.Errorf("Failed to fetch public key from %q: %v", jwksURI, err)
-		}
-		pubKey = string(resp)
-	}
-
-	r.keyEntries.Store(key, jwtPubKeyEntry{
-		pubKey:            pubKey,
-		lastRefreshedTime: now,
-		lastUsedTime:      now,
-	})
-
-	return pubKey, err
+	// fetching the public key in the background
+	jwksuriChannel <- key
+	return "", nil
 }
 
 // BuildLocalJwks builds local Jwks by fetching the Jwt Public Key from the URL passed if it is empty.
 func (r *JwksResolver) BuildLocalJwks(jwksURI, jwtIssuer, jwtPubKey string) *envoy_jwt.JwtProvider_LocalJwks {
 	if jwtPubKey == "" {
 		var err error
-
 		// jwtKeyResolver should never be nil since the function is only called in Discovery Server request processing
 		// workflow, where the JWT key resolver should have already been initialized on server creation.
 		jwtPubKey, err = r.GetPublicKey(jwtIssuer, jwksURI)
```

Comment:
With async resolve, isn't it always return an empty jwtPubKey?

Reply:
Only if the key is not present in the cache.

Reply:
I read it and seems it will not wait for async resolve result, do I miss something

Reply:
It won't wait instead we are pushing the key to the envoy in the backgroud.

Reply:
How does it push to envoy in the background?

Reply:
I am using the same logic as `refresher` which is always running in the background - https://github.com/istio/istio/blob/master/pilot/pkg/model/jwks_resolver.go#L488

Reply:
IC